osd: support window switcher with thumbnails

tokyo4j opened the pull request:
This PR implements window switcher with thumbnails (window previews):

https://github.com/user-attachments/assets/7a53934a-a1a2-4db9-bf58-3bd8f7421328

New configration `<windowSwitcher style="classic|thumbnail">` is added to switch between the old/new window switcher style.

Following theme entries are added:
```
osd.window-switcher.thumbnail.width.max: 80%
osd.window-switcher.thumbnail.padding: 4
osd.window-switcher.thumbnail.item.width: 300
osd.window-switcher.thumbnail.item.height: 250
osd.window-switcher.thumbnail.item.padding: 10
osd.window-switcher.thumbnail.item.active.border.width: 2
osd.window-switcher.thumbnail.item.active.border.color: #589bda
osd.window-switcher.thumbnail.item.active.bg.color: #c7e2fc
osd.window-switcher.thumbnail.item.icon.size: 60
```

Some theme entries for the old style are renamed with `osd.window-switcher.classic` prefix:
```
osd.window-switcher.classic.width: 600
osd.window-switcher.classic.padding: 4
osd.window-switcher.classic.item.padding.x: 10
osd.window-switcher.classic.item.padding.y: 1
osd.window-switcher.classic.item.active.border.width: 2
osd.window-switcher.classic.item.icon.size: 50
```

Following theme entries are unchanged and used by both old and new styles:
```
osd.bg.color: #e1dedb
osd.border.color: #000000
osd.border.width: 1
osd.label.text.color
```

TODO:
- More testing, especially with outputs and clients with transformation or scaling
- Fill documentations
- ~Find a way to get rid of `scene_buffer->WLR_PRIVATE.texture`, if possible? I'll need to learn APIs of wlroots more...~ I think we can just call `wlr_texture_from_buffer()`. It may incur some overhead compared to the scene renderer of wlroot as it always creates a new texture, but it will be negligible.

I'll appreciate testing or suggestions for the design.

Link: https://github.com/labwc/labwc/discussions/2172

jp7677 commented:
I gave this a short spin. Very beautiful and works nicely. A few initial thoughts:

- Some thumbnails (firefox, tillix or  ptyxis) appear to have a black bottom/right boarder (like those are not centered) when not in full screen state. I don't see this in e.g. steam, clion or gnome-terminal.
- with a lots of windows the thumbnails are divided into multiple rows (I really like this!). It would be very very cool if up/down arrow key would navigate between rows in that case instead of next/previous like arrow left/right key.
- With the thumbnail style, preview and outline options might be default to "no" instead of "yes" when not specified otherwise. But this is personal preference.

johanmalm commented:
Looks amazing. 

I get this crash:

```
Fontconfig warning: using without calling FcInit()
Failed to acquire service name: File exists
Is a notification daemon already running?
AddressSanitizer:DEADLYSIGNAL
=================================================================
==11051==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f56e7748334 bp 0x7fff3ffe74a0 sp 0x7fff3ffe7388 T0)
==11051==The signal is caused by a READ memory access.
==11051==Hint: address points to the zero page.
    #0 0x7f56e7748334 in wlr_buffer_get_dmabuf (/usr/lib/libwlroots-0.19.so+0x81334) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3)
    #1 0x7f56e7703c98  (/usr/lib/libwlroots-0.19.so+0x3cc98) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3)
    #2 0x5608e734a5c4 in render_node ../src/osd/osd-thumbnail.c:42
    #3 0x5608e734a301 in render_node ../src/osd/osd-thumbnail.c:35
    #4 0x5608e734a301 in render_node ../src/osd/osd-thumbnail.c:35
    #5 0x5608e734a301 in render_node ../src/osd/osd-thumbnail.c:35
    #6 0x5608e734b1b0 in render_thumb ../src/osd/osd-thumbnail.c:77
    #7 0x5608e734c496 in create_item_scene ../src/osd/osd-thumbnail.c:139
    #8 0x5608e734e28e in osd_thumbnail_create ../src/osd/osd-thumbnail.c:226
    #9 0x5608e7342969 in update_osd ../src/osd/osd.c:291
    #10 0x5608e73412d5 in osd_begin ../src/osd/osd.c:182
    #11 0x5608e71cff6d in run_action ../src/action.c:941
    #12 0x5608e71d2dcf in actions_run ../src/action.c:1395
    #13 0x5608e731b017 in handle_compositor_keybindings ../src/input/keyboard.c:565
    #14 0x5608e731c28c in handle_key ../src/input/keyboard.c:648
    #15 0x7f56e7f4e51d in wl_signal_emit_mutable (/usr/lib/libwayland-server.so.0+0x851d) (BuildId: 99ddf009eb5a6bbfb7d34377dd8c6a593ef0cd76)
    #16 0x7f56e77563c5 in wlr_keyboard_notify_key (/usr/lib/libwlroots-0.19.so+0x8f3c5) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3)
    #17 0x7f56e7720b76  (/usr/lib/libwlroots-0.19.so+0x59b76) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3)
    #18 0x7f56e6f44ac5  (/usr/lib/libffi.so.8+0x7ac5) (BuildId: 02fec367864097384b082accc299b6e74c83454b)
    #19 0x7f56e6f4176a  (/usr/lib/libffi.so.8+0x476a) (BuildId: 02fec367864097384b082accc299b6e74c83454b)
    #20 0x7f56e6f4406d in ffi_call (/usr/lib/libffi.so.8+0x706d) (BuildId: 02fec367864097384b082accc299b6e74c83454b)
    #21 0x7f56e688d90f  (/usr/lib/libwayland-client.so.0+0x490f) (BuildId: 081c8feac05f3afcca5e17942c83d162437d0953)
    #22 0x7f56e688e198  (/usr/lib/libwayland-client.so.0+0x5198) (BuildId: 081c8feac05f3afcca5e17942c83d162437d0953)
    #23 0x7f56e688e572 in wl_display_dispatch_queue_pending (/usr/lib/libwayland-client.so.0+0x5572) (BuildId: 081c8feac05f3afcca5e17942c83d162437d0953)
    #24 0x7f56e772096f  (/usr/lib/libwlroots-0.19.so+0x5996f) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3)
    #25 0x7f56e7f50111 in wl_event_loop_dispatch (/usr/lib/libwayland-server.so.0+0xa111) (BuildId: 99ddf009eb5a6bbfb7d34377dd8c6a593ef0cd76)
    #26 0x7f56e7f521f6 in wl_display_run (/usr/lib/libwayland-server.so.0+0xc1f6) (BuildId: 99ddf009eb5a6bbfb7d34377dd8c6a593ef0cd76)
    #27 0x5608e71f7b25 in main ../src/main.c:266
    #28 0x7f56e60376b4  (/usr/lib/libc.so.6+0x276b4) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
    #29 0x7f56e6037768 in __libc_start_main (/usr/lib/libc.so.6+0x27768) (BuildId: 468e3585c794491a48ea75fceb9e4d6b1464fc35)
    #30 0x5608e71c8dc4 in _start (/home/johan/tmp/labwc/build/labwc+0x2badc4) (BuildId: b1d26017eae689ea728abf039dff4a37ebd07cae)

==11051==Register values:
rax = 0x00007f56e77f1e01  rbx = 0x00007b56e0d0f8e0  rcx = 0x0000000000000000  rdx = 0x0000000000000000  
rdi = 0x0000000000000000  rsi = 0x00007fff3ffe7410  rbp = 0x00007fff3ffe74a0  rsp = 0x00007fff3ffe7388  
 r8 = 0x0000000000000000   r9 = 0x0000000000000014  r10 = 0x0000000000006000  r11 = 0x00000000000010f8  
r12 = 0x00007c86e1fe4d40  r13 = 0x00000f6adc1a1f00  r14 = 0x0000000000000000  r15 = 0x00007fff3ffe74f0  
AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/usr/lib/libwlroots-0.19.so+0x81334) (BuildId: fe0111da6b6ed06c643a5b2102d2d6e0ee9d00e3) in wlr_buffer_get_dmabuf
==11051==ABORTING
```

johanmalm commented:
Have not narrowed it down very much yet, but thought I'd just send what I can reproduce the crash with:

```
<?xml version="1.0"?>
<labwc_config>
  <windowSwitcher show="yes" preview="yes" outlines="yes" allWorkspaces="no">
    <style>thumbnail</style>
    <fields>
      <field content="type" width="20%"/>
      <field content="desktop_entry_name" width="30%"/>
      <field content="title" width="50%"/>
    </fields>
  </windowSwitcher>
  <keyboard>
    <default/>
    <keybind key="F12" name.action="ToggleKeybinds"/>
  </keyboard>
</labwc_config>
```

tokyo4j commented:
@johanmalm Can you try my updated branch? I added a guard against null `scene_buffer->buffer`.

johanmalm commented:
> @johanmalm Can you try my updated branch? I added a guard against null `scene_buffer->buffer`.

Seems to be fixed now.

johanmalm commented:
I have had a look at the code and it seems fine to me.

I like the osd-impl approach in that I think it enables us to strike the right balance between having osd modules with a small amount of configuration compared with having a "stupid" amount of theme options to cater for every wish.

Some specifics:

1. I feel we ought to consider a name other than `thumbnail` because: (a) with the modular nature introduced by this PR I think it's conceivable that we may accept other variants over time, some of which may contain thumbnails too; and (b) the word thumbnail makes some theme options non-intuitive. For example, it feels like `width.max` might refer to the thumbnail itself in `osd.window-switcher.thumbnail.width.max`.
  - Might we want to consider `s/classic/ob/` and `s/thumbnail/xf/` - as a recognition that we have drawn inspiration from those window managers in terms of the look and feel.
  - Or we could `s/thumbnail/tokyo/` :smile:
2. Consider handling window-switcher growing too big for the output. We don't currently deal with that, so might be a good candidate for a follow up PR.
3. Small (e.g. 50) values for `osd.window-switcher.thumbnail.item.{width,height}` make the thumbnails+icons disappear (presumably because the icon size is set to 60?)
4. With the addition of the `classic.` in some of theme keys, should we consider honoring backward compatibility? I suggest a couple of releases before deprecating.

iamloci commented:
Okay, I've been test driving this for the past week, and overall it looks great!

One note: it seems like the new thumbnail switcher doesn't respect the `<fields>` option. Is this intended? If so, will a follow-up PR address it? I'd like to still be able to display on which output/workspace a window is located.

Minimal example `rc.xml`:

```
<?xml version="1.0"?>
<labwc_config>
	<windowSwitcher>
		<style>thumbnail</style>
		<fields>
			<field content="workspace" width="30%" />
			<field content="title" width="70%" />
		</fields>
	</windowSwitcher>
</labwc_config>
```

jp7677 commented:
> Yeah I fixed that with https://github.com/labwc/labwc/commit/03c5e7286e6a3255fd7db39dbb1cd73c9effd209. Thanks for reporting.

You are welcome ofc!

> Using this for a while, I've also experienced 2 crashes, I think when activating the window switcher with a Steam full screen game. Sorry for being lazy and not really following up...

> Please report me if the crash still happens with my latest branch. It may be the same as what @johanmalm reported.

Both of your fixups where applied (I noticed because the black border issue was fixed). That said, I tried to reproduce just now, but I could not. Looking at `coredumpctl` though, the stack trace of those two crashes shows:

<details><summary>Stack trace</summary>
<p>

```
                #0  0x00000000004c6270 wlr_addon_find (/usr/bin/labwc + 0xc6270)
                #1  0x00000000004574b4 gles2_buffer_get_or_create (/usr/bin/labwc + 0x574b4)
                #2  0x0000000000457654 gles2_begin_buffer_pass (/usr/bin/labwc + 0x57654)
                #3  0x0000000000456e55 wlr_renderer_begin_buffer_pass (/usr/bin/labwc + 0x56e55)
                #4  0x000000000044dcb6 osd_thumbnail_create (/usr/bin/labwc + 0x4dcb6)
                #5  0x000000000044b9d6 update_osd (/usr/bin/labwc + 0x4b9d6)
                #6  0x000000000044bca0 osd_begin (/usr/bin/labwc + 0x4bca0)
                #7  0x000000000040fbd0 actions_run (/usr/bin/labwc + 0xfbd0)
                #8  0x0000000000445df5 handle_compositor_keybindings (/usr/bin/labwc + 0x45df5)
                #9  0x0000000000446324 handle_key (/usr/bin/labwc + 0x46324)
                #10 0x00007fed2a41031e wl_signal_emit_mutable (libwayland-server.so.0 + 0x431e)
                #11 0x00000000004ae3c0 wlr_keyboard_notify_key (/usr/bin/labwc + 0xae3c0)
                #12 0x0000000000473f49 handle_keyboard_key (/usr/bin/labwc + 0x73f49)
                #13 0x000000000047301b handle_libinput_readable (/usr/bin/labwc + 0x7301b)
                #14 0x00007fed2a411f42 wl_event_loop_dispatch (libwayland-server.so.0 + 0x5f42)
                #15 0x00007fed2a413fe5 wl_display_run (libwayland-server.so.0 + 0x7fe5)
                #16 0x000000000040e6fc main (/usr/bin/labwc + 0xe6fc)
                #17 0x00007fed29212575 __libc_start_call_main (libc.so.6 + 0x3575)
                #18 0x00007fed29212628 __libc_start_main@@GLIBC_2.34 (libc.so.6 + 0x3628)
                #19 0x000000000040e845 _start (/usr/bin/labwc + 0xe845)
``` 

</p>
</details> 

This was unfortunately a release build, but may be this is already of use. I'm now running a `debugoptimized` build, I'll report back when this happens again.

johanmalm commented:
@alpha6z Cool script. It may be simpler to write the equivalent within the compositor without too much bother. Suggest waiting until this PR is done and then take a view.

tokyo4j commented:
I've been considering what's the best name for the two styles... so I surveyed some popular window switcher implementations for reference.

### KDE Plasma

| Compact                                                                                                                                | Large icons                                                                                                                                | Thumbnail grid                                                                                                                                | Cover switch                                                                                                                                | Flip switch                                                                                                                                | Sidebar                                                                                                                                |
| -------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="1920" height="1080" alt="compact" src="https://github.com/user-attachments/assets/c96dce0d-3778-40a9-a467-008c2098f794" /> | <img width="1920" height="1080" alt="large icons" src="https://github.com/user-attachments/assets/5dae7025-ac09-405f-8eac-504e04e762b6" /> | <img width="1920" height="1080" alt="thumbnail grid" src="https://github.com/user-attachments/assets/be3362e3-1f8e-48f2-9d25-ab438ad44fc8" /> | <img width="1920" height="1080" alt="cover switch" src="https://github.com/user-attachments/assets/5a228aea-137e-4aed-83d5-16d4f7ba0fd9" /> | <img width="1920" height="1080" alt="flip switch" src="https://github.com/user-attachments/assets/8d665d67-fe49-40a3-b664-753e9bc64d5c" /> | <img width="1920" height="1080" alt="sidebar" src="https://github.com/user-attachments/assets/8b43c16f-15c7-4782-9ebc-79b019b3843c" /> |

### Openbox

| List                                                                                                                                | Icons                                                                                                                                |
| ----------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
| <img width="1920" height="1200" alt="list" src="https://github.com/user-attachments/assets/fe2aae75-3f85-4164-b627-328619626cf2" /> | <img width="1920" height="1200" alt="icons" src="https://github.com/user-attachments/assets/8e329b4a-6926-42b3-bd72-f625edef0d41" /> |

### Xfce4

| "Cycle through windows in a list" checked                                                                                                                     | "Cycle through windows in a list" unchecked                                                                                                                             |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <img width="1920" height="1200" alt="cycle-through-window-in-a-list" src="https://github.com/user-attachments/assets/b23e310e-0c58-4f35-9cbc-bb5d57843377" /> | <img width="1920" height="1200" alt="cycle-through-window-in-a-list-unchecked" src="https://github.com/user-attachments/assets/96403872-2acc-4f14-8aa3-edb7cf091c71" /> |

### [AltTab](https://alt-tab-macos.netlify.app/) for MacOS

| Titles                                                                                     | App icons                                                                                     | Thumbnails                                                                                     |
| ------------------------------------------------------------------------------------------ | --------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------- |
| ![Titles](https://github.com/user-attachments/assets/5fe0e821-97ae-4c11-a1a6-6fc8e127680e) | ![App icons](https://github.com/user-attachments/assets/f2d51e3c-ecf5-4bc3-82e2-11f71c6a20d3) | ![Thumbnails](https://github.com/user-attachments/assets/8986a077-c46d-40af-bd73-b214f569416f) |

> The reason I was trying to avoid the word thumbnail as a module/style name was that we may regret it in years to come if/when we consider merging other window-switcher modules with thumbnails.

I think it's unlikely that we will add another window-switcher module with thumbnails. As I showed above, the most popular styles are titles, icons+titles and thumbnails+icons+titles and I don't think there's much demand for anything else.

So I'd like to go with `(style-)classic` and `(style-)thumbnail`.

johanmalm commented:
Amazing research. Really interesting to see all those modes on the various DEs/OSs.

> So I'd like to go with `(style-)classic` and `(style-)thumbnail`.

I'm fine with that.

tokyo4j commented:
Rebased and added `style-` prefix to the new theme options.

This PR is now ready. Different scales/transformations of outputs/clients worked well in my test.

johanmalm commented:
This is great. Thanks for all the work with this.

LGTM. Have tested and it seems to work well.

tokyo4j commented:
> Hi! In this new style, the icons provided by the `xdg-toplevel-icon` protocol are not being displayed, showing only the fallback icon. As it is quite self-explanatory, I did not open an issue, but I can open one if you want.

Thanks. Opened #3033 to fix this.
